### PR TITLE
[Feature] Support SAGAN + Fix bugs in SNGAN

### PR DESCRIPTION
**Pull request by LeoXing1996**

### CODE REVIEW

**Comment by LeoXing1996** [comment]
Unit tests would be added after peer review.

File: mmgen/models/architectures/sngan_proj/generator_discriminator.py [hunk 1]
```diff
@@ -436,14 +479,30 @@ def __init__(self,
...
             self.from_rgb_cfg,
             dict(in_channels=input_channels, out_channels=base_channels))
 
         self.conv_blocks = nn.ModuleList()
-        for idx in range(len(downsample_cfg)):
+        # add self-attention block after the first block
+        if 1 in attention_after_nth_block:
+            attn_cfg_ = deepcopy(attention_cfg)
+            attn_cfg_['in_channels'] = base_channels
+            # attn_cfg_['out_channels'] = base_channels
```
Comment (nbei) [inline mmgen/models/architectures/sngan_proj/generator_discriminator.py:502]: If this line is dirty comment, you may remove it.

File: mmgen/models/architectures/sngan_proj/generator_discriminator.py [hunk 2]
```diff
@@ -436,14 +479,30 @@ def __init__(self,
...
                              '`channels_cfg`, but receive '
-                             f'{len(downsample_cfg)} and {len(channels_cfg)}.')
+                             f'{len(self.downsample_list)} and '
+                             f'{len(self.channel_factor_list)}.')
+
+        # check `attention_after_nth_block`
+        if not isinstance(attention_after_nth_block, list):
+            attention_after_nth_block = [attention_after_nth_block]
+        if not all([isinstance(idx, int)
+                    for idx in attention_after_nth_block]):
```
Comment (nbei) [inline mmgen/models/architectures/sngan_proj/generator_discriminator.py:489]: You may use `is_list_of` in mmcv:

```python
from mmcv.utils import is_list_of

if not is_list_of(attention_after_nth_block, int):
    raise xxx

```

File: mmgen/models/architectures/sngan_proj/generator_discriminator.py [hunk 3]
```diff
@@ -454,6 +513,13 @@ def __init__(self,
             block_cfg_['out_channels'] = factor_output * base_channels
             self.conv_blocks.append(build_module(block_cfg_))
 
+            # build self-attention block
+            if idx + 2 in attention_after_nth_block:
```
Comment (nbei) [inline mmgen/models/architectures/sngan_proj/generator_discriminator.py:517]: Please clarify why `+2` here.

File: mmgen/models/architectures/sngan_proj/generator_discriminator.py [hunk 4]
```diff
@@ -151,6 +171,14 @@ def __init__(self,
             block_cfg_['out_channels'] = factor_output * base_channels
             self.conv_blocks.append(build_module(block_cfg_))
 
+            # build self-attention block
+            if idx + 1 in attention_after_nth_block:
```
Comment (nbei) [inline mmgen/models/architectures/sngan_proj/generator_discriminator.py:175]: Please clarify why `+1` here.

File: mmgen/models/architectures/sngan_proj/generator_discriminator.py [hunk 5]
```diff
@@ -151,6 +171,14 @@ def __init__(self,
...
+                self.attention_block_idx = idx + 1
+                attn_cfg_ = deepcopy(attention_cfg)
+                attn_cfg_['in_channels'] = factor_output * base_channels
+                # attn_cfg_['out_channels'] = factor_output * base_channels
```
Comment (nbei) [inline mmgen/models/architectures/sngan_proj/generator_discriminator.py:179]: If this line is dirty comment, you can just remove it.

File: mmgen/models/architectures/sngan_proj/modules.py [hunk 6]
```diff
@@ -445,3 +451,97 @@ def init_weights(self):
             else:
                 constant_init(self.weight_embedding, 1)
                 constant_init(self.bias_embedding, 0)
+
+
+# TODO: retain or discard? rename or modify?
+@MODULES.register_module()
+class SelfAttentionBlock_SA(nn.Module):
```
[the remark this answers is not in the document]
Comment (LeoXing1996) [reply mmgen/models/architectures/sngan_proj/modules.py:458]: Would we need to support `out_channels` and other arguments in bigGAN's `SelfAttentionBlock`?

File: mmgen/models/architectures/sngan_proj/generator_discriminator.py [hunk 7]
```diff
@@ -277,19 +312,20 @@ def init_weights(self, pretrained=None, strict=True):
                         else:
                             xavier_init(
                                 m, gain=np.sqrt(2), distribution='uniform')
-                    elif isinstance(m, nn.Linear):
+                    if isinstance(m, nn.Linear):
                         xavier_init(m, gain=1, distribution='uniform')
-                    elif isinstance(m, nn.Embedding):
+                    if isinstance(m, nn.Embedding):
                         if 'weight' in n:
                             constant_init(m, 1)
```
Comment (nbei) [inline mmgen/models/architectures/sngan_proj/generator_discriminator.py:319]: Is this right?  You should check whether this implementation is right.

File: mmgen/models/architectures/sngan_proj/generator_discriminator.py (hunk 1 above)
Comment (nbei) [reply mmgen/models/architectures/sngan_proj/generator_discriminator.py:538]: Should this line be removed?

**Review by nbei — APPROVED** [review]

**Pull request merged**